script/commands/migrate.ts

```diff
@@ -0,0 +1,149 @@
+import { Command } from "commander";
+import { CodeDetails } from "@cosmjs/cosmwasm-stargate";
+
+import { CONTAINER, Dependencies } from "../shared/ioc";
+import { ContractNames } from "../shared/contract";
+import { ContextHook, ContextIsm } from "../shared/context";
+import { askQuestion, waitTx } from "../shared/utils";
+
+export const migrateCmd = new Command("migrate")
+  .description("Migrate contracts")
+  .configureHelp({ showGlobalOptions: true })
+  .option("-c --contracts <contracts...>", "specify contracts to migrate")
+  .action(handleMigrate);
+
+async function handleMigrate(_: any, cmd: any) {
+  const { contracts } = cmd.optsWithGlobals();
+
+  const { ctx, client } = CONTAINER.get(Dependencies);
+
+  const flatten = [
+    ...flattenIsm(ctx.deployments.isms),
+    ...flattenHook(ctx.deployments.hooks?.default),
+    ...flattenHook(ctx.deployments.hooks?.required),
+    ctx.deployments.core?.mailbox,
+    ctx.deployments.core?.validator_announce,
+    ctx.deployments.test?.msg_receiver,
+    ...(ctx.deployments.warp?.cw20 || []),
+    ...(ctx.deployments.warp?.native || []),
+  ].filter((v, i, arr) => arr.indexOf(v) === i) as {
+    type: ContractNames;
+    address: string;
+  }[];
+
+  const withContractInfo = await Promise.all(
+    (contracts
+      ? flatten.filter((v) => contracts.includes(v.type))
+      : flatten
+    ).map(async (v) => {
+      const contractInfo = await client.wasm.getContract(v.address);
+      const codeInfo = await client.wasm.getCodeDetails(contractInfo.codeId);
+      return { ...v, contractInfo, codeInfo };
+    })
+  );
+
+  const artifacts = Object.fromEntries(
+    await Promise.all(
+      Object.entries(ctx.artifacts).map(async ([contractName, codeId]) => {
+        const codeInfo = await client.wasm.getCodeDetails(codeId);
+        return [contractName, codeInfo];
+      })
+    )
+  ) as Record<ContractNames, CodeDetails>;
+
+  const toMigrate = withContractInfo.filter(
+    (v) =>
+      v.codeInfo.id !== ctx.artifacts[v.type] &&
+      v.codeInfo.checksum !== artifacts[v.type].checksum
+  );
+
+  if (toMigrate.length === 0) {
+    console.log("No changes detected.");
+    return;
+  }
+
+  if (!(await askQuestion("Do you want to proceed? (y/n)"))) {
```

Comment:
Should we log out what's gonna happen to allow the user to know?

Reply:
Agree!